pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -769,12 +784,21 @@ protected boolean commitSegment(String controllerVipUrl, boolean isSplitCommit)
     if (!segTarFile.exists()) {
       throw new RuntimeException("Segment file does not exist:" + segTarFileName);
     }
+    // Set the flag to true to prevent the server delete the segment for upload after receiving OFFLINE->ONLINE
+    // message from helix.
+    _waitingForUploadToSegmentStore = true;
     SegmentCompletionProtocol.Response commitResponse = commit(controllerVipUrl, isSplitCommit);
 
     if (!commitResponse.getStatus().equals(SegmentCompletionProtocol.ControllerResponseStatus.COMMIT_SUCCESS)) {
       return false;
     }
-
+    // Asynchronously upload the segment file to Pinot FS for backup. The upload result does not change the segment
+    // completion protocol flow.
+    if (!_indexLoadingConfig.isEnableSegmentUploadToController()) {
```

Comment:
perhaps the config is intended as isEnabledSegmentUploadToDeepStore()?)

Reply:
Yes. I should introduced one more flag to control the behavior. Right now my logic is that if server does not upload the segment to controller, it must upload the segment to deep storage -- which might not always be the case if there is no deep storage configured.

Reply:
Done. Check the segment store config too before doing upload.